Add support for s3 access points as opaque URLs

zeleena opened the pull request:
This is the second step after https://github.com/lyft/ignition/pull/1 where S3 access point ARN URLs are now use `scheme:path` and are valid opaque URIs.

If the S3 URL is opqaue and is an ARN, the bucket is the access point ARN ending and including the access point name, and the key is the remaining portion (path to S3 object). This follows the [access point](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-access-points.html) and [multi-region access point](https://docs.aws.amazon.com/AmazonS3/latest/userguide/MultiRegionAccessPointRequests.html) naming guides and is consist with how we use `aws s3api get-object --bucket <access point ARN> --key <path to object> <new file>`

On a Fedora node, tested with black box tests.
```
# ./build_blackbox_tests 
Using version from git...
Building ignition...
Building ignition-validate...
Compiling tests...
Building tests/stubs/groupdel-stub
Building tests/stubs/useradd-stub
Building tests/stubs/userdel-stub
Building tests/stubs/usermod-stub
Success
# sh -c 'PATH=$PWD/bin/amd64:$PATH ./tests.test'
PASS
```
Manually ran go tests that showed it would split `u.Opaque` which is `arn:aws:s3::123456789012:accesspoint/abcdefghijklm.mrap/resources/anonymous` into
```
bucket = arn:aws:s3::123456789012:accesspoint/abcdefghijklm.mrap
key = resources/anonymous
```
It currently only supports a `/` before the access point name and not a `:`

Internally at Lyft I'll do manual testing by building custom ignition and trying to use multi-region access points with it.

zeleena committed:
Migrate to aws-sdk-go-v2 with changes in aws.go and resource/url.go

zeleena committed:
Ran 'make vendor' to remove v1 and get v2 modules

zeleena committed:
Use context.Background and fix the way downloader uses anonymous credentials

zeleena committed:
Run 'make vendor' for latest change

zeleena committed:
Use regionHint when getting bucket region and simplify

zeleena committed:
Reran 'make vendor' for last change

zeleena committed:
Add support for s3 access points as opaque URLs

paulnivin commented on internal/resource/url.go:
```diff
@@ -408,6 +409,18 @@ func (f *Fetcher) fetchFromS3(u url.URL, dest s3target, opts FetchOptions) error
+	// Determine if the s3 URI is an access point ARN
+	bucket := u.Host
+	key := u.Path
+	if arn.IsARN(u.Opaque) {
```
recommend doing "if Opaque" { ... } else { bucket = u.Host
key = u.Path }" so the vars are set once

zeleena replied:
done

zeleena committed:
Set bucket and key only once

zeleena committed:
Trim s3 key path and only get bucket region for non access point urls

zeleena committed:
Add region to s3client to prevent download error

sohankunkerkar committed:
internal/*: change the location of Ignition report

Fixes https://github.com/coreos/fedora-coreos-tracker/issues/977

This change helps to address the problem of displaying a warning
about running Ignition twice while reprovisioning a bare metal
CoreOS instance with a persistent /var partition.

jlebon committed:
internal/exec/util: rename FindFirstMissingDirForFile and tweak docs

It actually returns a path, not a component. Also, it can return the
path itself if just the final component is missing or if it already
exists. Let's tweak the docstring and use the term "component" in its
name to be clearer.

travier committed:
docs: Do not convert -- & --- to en/em-dash

'--' is frequently used for command line options and was thus
incorrectly rendered as a special en-dash symbol.

travier committed:
docs: Remove default layout from front matter

No need to specify it as it is the default.

jlebon committed:
Merge pull request #1270 from travier/docfixes

jlebon committed:
Merge pull request #1268 from jlebon/pr/tweak-first-missing-name

bgilbert committed:
docs/supported-platforms: add platform IDs

bgilbert committed:
docs/supported-platforms: sort by platform ID

bgilbert committed:
docs/supported-platforms: update platform names and URLs

bgilbert committed:
docs/supported-platforms: add missing platforms

Omit VirtualBox for now, since our current userdata mechanism there isn't
very usable.

Fixes https://github.com/coreos/ignition/issues/1222.

bgilbert committed:
test: ensure all platforms are documented

Skip VirtualBox for now, since our current userdata mechanism there isn't
very usable.  Omit "file" since it's a test platform.

bgilbert committed:
docs/supported-platforms: drop reference to platform-specific agents

This is a holdover from Container Linux.

bgilbert committed:
docs/supported-platforms: switch to Afterburn docs URL

sohankunkerkar committed:
Merge pull request #1267 from sohankunkerkar/fix-ignition-prev-run

internal/*: change the location of Ignition report

jlebon committed:
stages/files: make variable name follow Go convention

My editor told me "don't use underscores in Go names".

jlebon committed:
stages/files: rename `relabelDirsForFile` and add docstring

Follow-up to #1268.

bgilbert committed:
Merge pull request #1271 from bgilbert/platforms

docs/supported-platforms: update for current platform list

jlebon committed:
Merge pull request #1272 from jlebon/pr/more-name-tweaking

Commit:
stages/filesystems: use mkfs.fat instead of mkfs.vfat

Ignition should use mkfs.fat instead of mkfs.vfat because mkfs.vfat is
symlink to mkfs.fat and may be absent in some distributions.
Dosfstools always creates mkfs.fat, but mkfs.vfat is optional
(https://github.com/dosfstools/dosfstools/blob/master/src/Makefile.am)

bgilbert committed:
Merge pull request #1273 from keremet/main

stages/filesystems: use mkfs.fat instead of mkfs.vfat

dghubble reviewed (commented):
close?

dghubble reviewed (changes requested):
trapped as a reviewer
